process-controller/src/main/java/org/jboss/as/process/jdk/JdkType.java

```diff
@@ -0,0 +1,151 @@
+    private static final String JAVA_HOME_ENV_VAR = "JAVA_HOME";
+    private static final String OS_NAME_SYS_PROP = "os.name";
+    private static final Collection<String> DEFAULT_MODULAR_JDK_ARGUMENTS;
+    private static final String JAVA_EXECUTABLE;
+    private static final String JAVA_UNIX_EXECUTABLE = "java";
+    private static final String JAVA_WIN_EXECUTABLE = "java.exe";
+    private final boolean isModularJdk;
+    private final String javaHome;
+    private final String javaExecutable;
+
+    static {
+        final String os = SecurityUtils.getSystemProperty(OS_NAME_SYS_PROP).toLowerCase(Locale.ROOT);
+        JAVA_EXECUTABLE = os.contains("win") ? "java.exe" : "java";
+
+        final ArrayList<String> modularJavaOpts = new ArrayList<>();
+        modularJavaOpts.add("--add-exports=java.base/sun.nio.ch=ALL-UNNAMED");
+        modularJavaOpts.add("--add-exports=jdk.unsupported/sun.reflect=ALL-UNNAMED");
+        modularJavaOpts.add("--add-exports=jdk.unsupported/sun.misc=ALL-UNNAMED");
+        modularJavaOpts.add("--add-modules=java.se");
+        DEFAULT_MODULAR_JDK_ARGUMENTS = Collections.unmodifiableList(modularJavaOpts);
+    }
+
+    private JdkType(final boolean isModularJdk, final String javaHome, final String javaExecutable) {
+        this.isModularJdk = isModularJdk;
+        this.javaHome = javaHome;
+        this.javaExecutable = javaExecutable;
+    }
+
+    public String getJavaHome() {
+        return javaHome;
+    }
+
+    public String getJavaExecutable() {
+        return javaExecutable;
+    }
+
+    public boolean isModularJdk() {
+        return isModularJdk;
+    }
+
+    public Collection<String> getDefaultArguments() {
+        return isModularJdk ? DEFAULT_MODULAR_JDK_ARGUMENTS : Collections.EMPTY_LIST;
+    }
+
+    public static JdkType createFromJavaHome(final File javaHome) {
+        return javaHome == null ? createFromJavaHome((String) null) : createFromJavaHome(javaHome.getAbsolutePath());
+    }
+
+    public static JdkType createFromJavaHome(final String javaHome) {
+        if (javaHome == null || javaHome.trim().equals("")) {
+            throw ProcessLogger.ROOT_LOGGER.invalidJavaHome(javaHome);
+        }
+        final File javaHomeDir = new File(javaHome);
+        if (!javaHomeDir.exists()) {
+            throw ProcessLogger.ROOT_LOGGER.invalidJavaHome(javaHomeDir.getAbsolutePath());
+        }
+        final File javaBinDir = new File(javaHomeDir, BIN_DIR);
+        if (!javaBinDir.exists()) {
+            throw ProcessLogger.ROOT_LOGGER.invalidJavaHomeBin(javaBinDir.getAbsolutePath(), javaHomeDir.getAbsolutePath());
+        }
+        final File javaExecutable = new File(javaBinDir, JAVA_EXECUTABLE);
+        if (!javaExecutable.exists()) {
+            throw ProcessLogger.ROOT_LOGGER.cannotFindJavaExe(javaBinDir.getAbsolutePath());
+        }
+        return new JdkType(new File(javaHomeDir, JMODS_DIR).exists(), javaHomeDir.getAbsolutePath(), javaExecutable.getAbsolutePath());
```

Comment:
Could you look into (if you haven't already) whether this FS based detection scheme is reliable for other JVMs we support (IBM etc), and if there is any variance on where this directory might exist (e.g. installation options someone might pick).

Reply:
I'm concerned that what we're doing here is imposing constraints on how people configure things.

For example, what if --default-jvm=/usr/bin/java or something like that? Or /bin/java or /lib/java (both of which exist on my laptop).  In the existing code, we're not imposing constraints on the value of --default-jvm, but in this PR we do because JdkType.createFromJavaExecutable wants to walk from its value up two levels in order to find "java home" and then look for jmods under it.

The --default-jvm value is user controllable in domain.sh via the $JAVA env var.

Reply:
Ok this is what I was worried about, that the layout is unreliable. I think the solution for standalone.* should be mirrored in domain mode. Before launching the real JVM, a leading detection execution using the --add-module + -version error code trick could be performed.